_posts/2017-02-21-Rails-5-1-beta1.markdown

```diff
@@ -0,0 +1,108 @@
+
+These tests are not without trade-offs. It's of course still slower to run through a whole browser setup than just test a model with a stubed out database. But it also tests so much more. You'd do well to familiarize yourself with system tests and have them as part of your testing answer.
+
+Thanks to [Eileen M. Uchitelle](https://github.com/eileencodes) for her work extracting this from [Basecamp](https://basecamp.com)!
+
+
+## Encrypted secrets
+
+If you're checking production passwords, api keys, and other secrets undisguised into your revision control system, you're doing it wrong. That's not safe and you should stop it! Now that's an easy prescription, but without a coherent answer to what you should do instead, it's also not that helpful.
+
+People have long been loading up the ENV to store these secrets or used a variety of other solutions. There are all sorts of trade-offs and drawbacks to the ENV-model, not least of which that you _still_ need to store those secrets for real somewhere else.
+
+Inspired by [Ara T. Howard](https://github.com/ahoward)'s [sekrets gem](https://github.com/ahoward/sekrets), we've [built encrypted secrets management into Rails 5.1](https://github.com/rails/rails/pull/28038). You can setup a new encrypted secrets file with `bin/rails secrets:setup`. That'll generate a master key you'll store outside of the repository, but allow you to commit the actual production secrets to your revision control. They're then decrypted in production either through an injected key file or through RAILS_MASTER_KEY in the ENV.
+
+Thank you to [Kasper Timm Hansen][kasper] for the work on this and Ara for the inspiration!
+
+
+## Parameterized mailers
+
+Action Mailer is modeled on Action Controller. It shares underpinnings through Action Dispatch, but it's long been disadvantaged from its controller cousin in the way it can share logic between actions. 
+
+In Action Controller, it's common to use before_action and similar callbacks to extract logic that applies to multiple actions. This is doable because the params hash is available before the action is invoked. But in Action Mailer, we've been using regular method signatures with explicit arguments, so those arguments haven't been available to filters that run before the actions.
+
+With [Parameterized Mailers](https://github.com/rails/rails/blob/master/actionmailer/lib/action_mailer/parameterized.rb#L2), we now give you the option of calling mailers with parameters that, like in controllers, are available before the action is invoked. This combines with the default to/from/reply_to headers to dramatically DRY-up some mailer actions.
```

Comment:
It is good to point to a specific commit because they files can move in the future.